pkg/runtime/controller.go

```diff
@@ -0,0 +1,399 @@
+			c.configState[ev.Key] = ev.Value
+		case store.Delete:
+			delete(c.configState, ev.Key)
+		}
+	}
+	c.publishSnapShot()
+}
+
+// filterInstanceConfig filters the configState and produces instanceConfig
+// that points to valid templates.
+func (c *Controller) filterInstanceConfig() map[string]*cpb.Instance {
+	instanceConfig := make(map[string]*cpb.Instance)
+
+	// first pass get all the validated instance and handler references
+	for k, cfg := range c.configState {
+		if _, found := c.templateInfo[k.Kind]; !found {
+			continue
+		}
+		instanceConfig[k.String()] = &cpb.Instance{
+			Name:     k.Name,
+			Template: k.Kind,
+			Params:   cfg,
+		}
+	}
+	return instanceConfig
+}
+
+// filterHandlerConfig filters the configState and produces handlerConfig
+// that points to valid adapters.
+func (c *Controller) filterHandlerConfig() map[string]*cpb.Handler {
+	handlerConfig := make(map[string]*cpb.Handler)
+	for k, cfg := range c.configState {
+		if _, found := c.adapterInfo[k.Kind]; !found {
+			continue
+		}
+		handlerConfig[k.String()] = &cpb.Handler{
+			Name:    k.Name,
+			Adapter: k.Kind,
+			Params:  cfg,
+		}
+	}
+	return handlerConfig
+}
+
+// convertToRuntimeRules converts internal rules to the format that resolver needs.
+func convertToRuntimeRules(ruleConfig map[string]map[string]*Rule) map[string][]*Rule {
+	// convert rules
+	rules := make(map[string][]*Rule)
+	for ns, nsmap := range ruleConfig {
+		rulesArr := make([]*Rule, 0, len(nsmap))
+		for _, rule := range nsmap {
+			rulesArr = append(rulesArr, rule)
+		}
+		rules[ns] = rulesArr
+	}
+	return rules
+}
+
+// processRules builds the current consistent view of the rules keyed by Namespace and then Name.
+// ht (handlerTable) keeps track of handler-instance association.
+func (c *Controller) processRules(handlerConfig map[string]*cpb.Handler,
+	instanceConfig map[string]*cpb.Instance, ht *handlerTable) map[string]map[string]*Rule {
+	// current consistent view of the rules
+	// keyed by Namespace and then Name.
+	ruleConfig := make(map[string]map[string]*Rule)
+
+	// check rules and ensure only good handlers and instances are used.
+	// record handler - instance associations
+	for k, cfg := range c.configState {
+		if k.Kind != RulesKind {
+			continue
+		}
+		rulec := cfg.(*cpb.Rule)
+		rule := &Rule{
+			selector: rulec.Selector,
+			name:     k.Name,
+		}
+		acts := c.processActions(rulec.Actions, handlerConfig, instanceConfig, ht)
+
+		rn := ruleConfig[k.Namespace]
+		if rn == nil {
+			rn = make(map[string]*Rule)
+			ruleConfig[k.Namespace] = rn
+		}
+		ruleActions := make(map[adptTmpl.TemplateVariety][]*Action)
+		for vr, amap := range acts {
+			for _, cf := range amap {
+				ruleActions[vr] = append(ruleActions[vr], cf)
+			}
+		}
+		rule.actions = ruleActions
+		rn[k.Name] = rule
+	}
+
+	return ruleConfig
+}
+
+const cleanupSleepTime = 500 * time.Millisecond
+
+// processActions prunes actions that lack referential integrity and associate instances with
```

Comment:
Are we not doing ref identity check at all ? The classic vaidator should be running before this code right ?